File: algorithms/0496.next_greater_element_i/next_greater_element_i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> greater_nums(nums1.size(), -1);
        stack<int> mono_stack;
        unordered_map<int, int> nums1_map;
        for (int i = 0; i < nums1.size(); ++i) {
            nums1_map[nums1[i]] = i;
        }
        for (int i = 0; i < nums2.size(); ++i) {
            const int &num = nums2[i];
            while (!mono_stack.empty() && mono_stack.top() < num) {
                const int prev_index = nums1_map[mono_stack.top()]; mono_stack.pop();
                greater_nums[prev_index] = num;
            }
            if (nums1_map.find(num) != nums1_map.end()) {
                mono_stack.push(num);
            }
        }
        return greater_nums;
    }
};

File: algorithms/0496.next_greater_element_i/next_greater_element_i.rs
use std::collections::HashMap;

impl Solution {
    pub fn next_greater_element(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut ans = vec![-1; nums1.len()];
        let mut stack = Vec::<i32>::new();
        let mut nums1_map = HashMap::new();
        for (i, &n) in nums1.iter().enumerate() {
            nums1_map.insert(n, i);
        }
        for &cur in nums2.iter() {
            while !stack.is_empty() && *(stack.last().unwrap()) < cur {
                let prev = nums1_map[stack.last().unwrap()];
                stack.pop();
                ans[prev] = cur;
            }
            if nums1_map.get(&cur).is_some() {
                stack.push(cur);
            }
        }
        ans
    }
}